File: src/main/scala/apvd/lib/Ellipse.scala
package apvd.lib

import apvd.lib.Ellipse.toDegrees

import scala.math.{ Pi, atan, cos, sin, sqrt }

sealed trait Ellipse {
  def color: String
  def name: String

  def cx: Double
  def cy: Double
  def rx: Double
  def ry: Double
  def theta: Double

  // Coefficients of this ellipse when represented as the locus of points satisfying Ax² + Bxy + Cy² + Dx + Ey + F = 0
  def A: Double
  def B: Double
  def C: Double
  def D: Double
  def E: Double
  def F: Double

  lazy val c = cos(theta)
  lazy val s = sin(theta)

  lazy val (rM, rm) = if (rx > ry) (rx, ry) else (ry, rx)

  lazy val fd = sqrt(rM*rM - rm*rm)
  lazy val fr = fd / rM

  lazy val xdir = Point(c, s)
  lazy val ydir = Point(-s, c)
  lazy val fdir = if (rx > ry) xdir else ydir

  lazy val (vx1, vx2) = center ± (xdir * rx)

  lazy val (vy1, vy2) = center ± (ydir * ry)

  lazy val ( f1,  f2) = center ± (fdir * fd)

  def center = Point(cx, cy)

  lazy val degrees = toDegrees(theta)

  def moveCenter(dx: Double, dy: Double): Ellipse =
    Coords(cx + dx, cy + dy, rx, ry, theta, color, name)

  import Ellipse.angle

  def moveVx(dx: Double, dy: Double): Ellipse = {
    val vx = vx1 + (dx, dy)
    val d = vx - center
    val rx = d.magnitude
    Coords(
      cx, cy,
      rx, ry,
      angle(d),
      color,
      name
    )
  }

  def moveVy(dx: Double, dy: Double): Ellipse = {
    val vy = vy1 + (dx, dy)
    val d = vy - center
    val ry = d.magnitude
    Coords(
      cx, cy,
      rx, ry,
      angle(d) - Pi/2,
      color,
      name
    )
  }

  def moveFocus(dx: Double, dy: Double): Ellipse = {
    val f = f1 + (dx, dy)
    val fd = center - f
    val fl = fd.magnitude
    val rM = sqrt(rm*rm + fl*fl)
    val (rX, rY) = if (rx >= ry) (rM, rm) else (rm, rM)
    val theta = angle(fd) + (if (rx >= ry) 0 else -Pi/2)
    Coords(
      cx, cy,
      rX, rY,
      theta,
      color,
      name
    )
  }

  /**
   * Express this ellipse in "coordinate" form (center x/y, radius x/y, CCW rotation)
   */
  def toCoords: Coords =
    Coords(
      cx = cx,
      cy = cy,
      rx = rx,
      ry = ry,
      theta = theta,
      color = color,
      name = name
    )

  /**
   * Express this ellipse in "coefficient" form (Ax² + Bxy + Cy² + Dx + Ey + F = 0)
   */
  def toCoeffs: Coeffs =
    Coeffs(
      A = A,
      B = B,
      C = C,
      D = D,
      E = E,
      F = F,
      color = color,
      name = name
    )

  /**
   * Projection that would turn this ellipse into the unit circle.
   */
  lazy val projection =
    Transforms(
      Seq(
        Translate(-cx, -cy),
        Rotate(-theta),
        Scale(1 / rx, 1 / ry)
      )
    )

  lazy val inversion = projection.invert

  def invert(p: Point): Point = p(inversion)

  def project(p: Point): Point =
    p.-(cx, cy).rotate(-theta).*(1 / rx, 1 / ry)

  def apply(transform: Option[Transform]): Ellipse = transform.map(apply).getOrElse(this)
  def apply(transform: Transform): Ellipse =
    transform match {
      case Translate(x, y) ⇒ translate(x, y)
      case Rotate(theta) ⇒ rotate(theta)
      case Scale(x, y) ⇒ scale(x, y)
      case Transforms(transforms) ⇒ transforms.foldLeft(this)(_ apply _)
    }

  /**
   * Translate, rotate, and scale this ellipse
   */
  def affine(sx: Double,
             sy: Double,
             theta: Double,
             tx: Double,
             ty: Double): Ellipse =
    translate(tx, ty)
      .rotate(theta)
      .scale(sx, sy)

  def translate(tx: Double, ty: Double): Ellipse
  def rotate(theta: Double): Ellipse
  def scale(sx: Double, sy: Double): Ellipse
}

case class Coords(cx: Double,
                  cy: Double,
                  rx: Double,
                  ry: Double,
                  theta: Double,
                  color: String,
                  name: String)
  extends Ellipse {

  lazy val rx2 = rx * rx
  lazy val ry2 = ry * ry
  lazy val c2 = c*c
  lazy val s2 = s*s
  lazy val cs = c*s

  lazy val d1 = ry * (cx * c + cy * s)
  lazy val d2 = rx * (cy * c - cx * s)
  lazy val d = rx2 * ry2 - d1 * d1 - d2 * d2

  lazy val a1 = c2 * ry2 + s2 * rx2
  lazy val c1 = c2 * rx2 + s2 * ry2

  lazy val r1 = ry2 - rx2

  lazy val A = a1
  lazy val B = 2*c*s*r1
  lazy val C = c1
  lazy val D = -2 * (cx*a1 + cy*c*s*r1)
  lazy val E = -2 * (cy*c1 + cx*c*s*r1)
  lazy val F = -d

  override def translate(tx: Double, ty: Double): Ellipse =
    copy(
      cx = cx + tx,
      cy = cy + ty
    )

  override def rotate(theta: Double): Ellipse = {
    val Point(cx, cy) = center.rotate(theta)
    copy(
      cx = cx,
      cy = cy,
      theta = this.theta + theta
    )
  }

  override def scale(sx: Double, sy: Double): Ellipse = toCoeffs.scale(sx, sy)
}

case class Coeffs(A: Double,
                  B: Double,
                  C: Double,
                  D: Double,
                  E: Double,
                  F: Double,
                  color: String,
                  name: String)
  extends Ellipse {
  /**
   * TODO: a given (A,B,C,D,E,F) ellipse's "rx" and "ry" vectors can be interpreted as being in any of 4 "cardinal"
   * directions.
   *
   * The [[theta]] below will be the one in [0,π/2), but this creates a possibility for discontinuous jumps in
   * representations of similar ellipses that straddle a boundary where e.g. [[theta]] jumps from π/2-ε to π/2+ε and
   * then is "warped" to ε, with corresponding swaps in the derived "x"- and "y"-values (e.g. [[rx]]/[[ry]],
   * {[[vx1]],[[vx2]]}/{[[vy1]],[[vy2]]}).
   *
   * This isn't currently causing issues as the marginal-ellipse-updating methods ([[moveFocus]], [[moveVx]],
   * [[moveVy]]) emit [[Coords]] and so [[Coeffs]] representations aren't stressed in ways that reveal such artifacts.
   *
   * One manifestation is that [[Coords]] whose [[theta]] is outside the interval [0,π/2) will be effectively rotated
   * 90º, 180º, or 270º when round-tripped through [[Coeffs]] (e.g. using [[toCoeffs]] followed by [[toCoords]]).
   *
   * Longer term, a [[theta]] be added to the [[Coeffs]] ctor to disambiguate the derived [[rx]]/[[ry]]/[[vx1]]/etc.
   * members, or the distinct "vx"/"vy" vertex-fields should be blurred into a ring of four vertices without a
   * prescribed notion of "x"- and "y"-axes.
   */
  lazy val theta =
    if (B == 0)
      0
    else
      atan(B / (A - C)) / 2

  lazy val c2 = c * c
  lazy val s2 = s * s
  lazy val cs = c * s

  lazy val leveled = this.rotate(-theta)

  lazy val Coeffs(nA, nB, nC, nD, nE, _, _, _) = leveled

  lazy val Point(cx, cy) = Point(-nD / nA / 2, -nE / nC / 2).rotate(theta)

  lazy val (rx, ry) = {
    val discrim = nD * nD / nA / 4 + nE * nE / nC / 4 - F
    (sqrt(discrim / nA), sqrt(discrim / nC))
  }

  override def translate(tx: Double, ty: Double): Coeffs =
    copy(
      D = D - 2*A*tx - B*ty,
      E = E - 2*C*ty - B*tx,
      F = F + A*tx*tx + B*tx*ty + C*ty*ty - D*tx - E*ty
    )

  override def rotate(theta: Double): Coeffs = {
    val (c, s) = (cos(theta), sin(theta))
    val (c2, cs, s2) = (c*c, c*s, s*s)
    copy(
      A = A*c2 - B*cs + C*s2,
      B = 2*cs*(A - C) + B*(c2 - s2),
      C = A*s2 + B*cs + C*c2,
      D = D*c - E*s,
      E = D*s + E*c
    )
  }

  override def scale(sx: Double, sy: Double): Coeffs =
    copy(
      A = A / sx / sx,
      B = B / sx / sy,
      C = C / sy / sy,
      D = D / sx,
      E = E / sy
    )
}

object Ellipse {
  def toDegrees(theta: Double): Double = theta * 180 / Pi
  def toTheta(degrees: Double): Double = degrees * Pi / 180

  def apply(cx: Double,
            cy: Double,
            rx: Double,
            ry: Double,
            theta: Double,
            color: String,
            name: String): Ellipse =
    Coords(cx, cy, rx, ry, theta, color, name)

  def angle(d: Point): Double = angle(d.x, d.y)
  def angle(dx: Double, dy: Double): Double =
    (
      if (dx == 0)
        if (dy < 0)
          -Pi / 2
        else
          Pi / 2
      else
        atan(dy / dx)
    ) + (
      (dx < 0, dy < 0) match {
        case (true, true) ⇒ Pi
        case (_, true) ⇒ 2*Pi
        case (true, false) ⇒ Pi
        case _ ⇒ 0
      }
    )
}

File: src/lib/ellipse.ts
import Intersection from './intersection';
import {pi, r3, sqrt} from './math';
import {quartic} from '../quartic';
import {pp, zeroCheck} from './utils';
import {Point} from "../components/point";

export type XY = [ number, number ];
export type Polar = { r: number, t: number }
export type Center = { cx: number, cy: number }
export type RadiusVector = { theta: number } & ({ rx: number } | { ry: number })

export default class Ellipse {
    idx: number;
    name: string;
    A: number;
    B: number;
    C: number;
    D: number;
    E: number;
    F: number;
    cx: number;
    cy: number;
    rx: number;
    ry: number;
    theta: number;
    color: string;
    degrees: number;
    cos: number;
    sin: number;
    rM: number;
    rm: number;
    fd: number;
    vx1: Point;
    vy1: Point;
    vx2: Point;
    vy2: Point;
    f1: Point;
    f2: Point;

    constructor(o: { [k: string]: any }) {
        this.idx = o.idx;
        this.name = o.name;
        let theta: number | undefined, degrees: number | undefined;
        if ('A' in o) {
            if (isNaN(o.A)) {
                throw new Error("Bad ellipse ctor: " + JSON.stringify(o));
            }
            this.A = o.A;
            this.B = o.B;
            this.C = o.C;
            this.D = o.D;
            this.E = o.E;
            this.F = o.F;
            theta = o.theta;
            if (o.C !== o.A) {
                theta = this.theta = Math.atan(o.B / (o.A - o.C)) / 2;
            }
        }

        if (!('cx' in o)) {
            if (theta === undefined) {
                throw new Error("expected t in Ellipse ctor: " + JSON.stringify(o));
            }
            const c = Math.cos(theta);
            const s = Math.sin(theta);

            if (Math.abs(o.B) < 1e-10) {
                const {A, C, D, E, F} = o;
                this.cx = -D / 2 / A;
                this.cy = -E / 2 / C;
                const n = -4 * F + D * D / A + E * E / C;
                this.rx = sqrt(n / A) / 2;
                this.ry = sqrt(n / C) / 2;

            } else {
                const e = this.rotate(-theta);
                const { cx, cy } = e;
                this.rx = e.rx;
                this.ry = e.ry;
                this.cx = c*cx - s*cy;
                this.cy = s*cx + c*cy;
            }
        } else {
            this.cx = o.cx;
            this.cy = o.cy;
            this.rx = o.rx;
            this.ry = o.ry;
            if (o.degrees !== undefined) {
                degrees = o.degrees;
            } else {
                theta = o.theta
            }
        }

        this.color = o.color;

        if (theta === undefined) {
            if (degrees === undefined) {
                throw new Error("expected degrees in Ellipse ctor: " + JSON.stringify(o));
            }
            theta = degrees * pi / 180;
        }
        if (degrees === undefined) {
            if (theta === undefined) {
                throw new Error("expected theta in Ellipse ctor: " + JSON.stringify(o));
            }
            degrees = 180 * theta / pi;
        }
        this.theta = theta
        this.degrees = degrees

        this.cos = Math.cos(this.theta);
        this.sin = Math.sin(this.theta);

        const { cx, cy, rx, ry, cos, sin, } = this;

        if (!('A' in o)) {
            const rx2 = rx * rx;
            const ry2 = ry * ry;
            const c = cos;
            const s = sin;
            const c2 = c*c;
            const s2 = s*s;

            const d1 = ry * (cx * c + cy * s);
            const d2 = rx * (cy * c - cx * s);
            let d = rx2 * ry2 - d1 * d1 - d2 * d2;

            const a1 = c2 * ry2 + s2 * rx2;
            const c1 = c2 * rx2 + s2 * ry2;

            const r1 = ry2 - rx2;

            this.A = a1;
            this.B = 2*c*s*r1;
            this.C = c1;
            this.D = -2 * (cx*a1 + cy*c*s*r1);
            this.E = -2 * (cy*c1 + cx*c*s*r1);
            this.F = -d;
        } else {
            this.A = o.A;
            this.B = o.B;
            this.C = o.C;
            this.D = o.D;
            this.E = o.E;
            this.F = o.F;
        }

        this.rM = Math.max(rx, ry);
        this.rm = Math.min(rx, ry);

        this.fd = sqrt(this.rM*this.rM - this.rm*this.rm);
        const fr = this.fd / this.rM;

        this.vx1 = { x: cx + rx*cos, y: cy + rx*sin }
        this.vy1 = { x: cx - ry*sin, y: cy + ry*cos }

        this.vx2 = { x: 2*cx - this.vx1.x, y: 2*cy - this.vx1.y }
        this.vy2 = { x: 2*cx - this.vy1.x, y: 2*cy - this.vy1.y }

        this.f1 ={
            x: cx + fr*(((rx >= ry) ? this.vx1.x : this.vy1.x) - cx),
            y: cy + fr*(((rx >= ry) ? this.vx1.y : this.vy1.y) - cy),
        };
        this.f2 = {
            x: cx - fr*(((rx >= ry) ? this.vx1.x : this.vy1.x) - cx),
            y: cy - fr*(((rx >= ry) ? this.vx1.y : this.vy1.y) - cy),
        }
    }

    toString() {
        return "E(c:" + pp(this.cx, this.cy) +
            "; r:" + pp(this.rx, this.ry) + "," +
            r3(this.degrees) +
            "°,(" + [this.A, this.B, this.C, this.D, this.E, this.F].map(r3).join(",") + "))";
    }

    s() {
        return this.toString();
    }

    modify(o: Center | RadiusVector): Ellipse {
        const { rx, ry, theta, cx, cy, color, idx, name } = this;
        return new Ellipse({ ...{ rx, ry, theta, cx, cy, color, idx, name }, ...o });
    }

    polar(x: number, y: number): Polar {
        const p = this.transform(x, y);
        const r = sqrt(p[0] * p[0] + p[1] * p[1]);
        //console.log("transformed:", pp(x,y), "→", pp(p), "r:", r, "p[0]/r:", p[0]/r);
        if (r === 0) return { r: r, t: 0 };
        // var atan = (p[0] == 0) ?
        const t = (p[1] < 0) ? -Math.acos(p[0] / r) : Math.acos(p[0] / r);
        //console.log("polar:", pp(x,y), pp(r,t));
        return { r: r, t: t };
    }

    transform(x: number | XY | Point, y?: number): XY {
        if (x instanceof Array) {
            y = x[1];
            x = x[0];
        } else if (typeof x === 'object' && 'x' in x) {
            y = x.y;
            x = x.x;
        } else if (y === undefined) {
            throw new Error("expected y in Ellipse.transform: " + JSON.stringify(x));
        }
        x -= this.cx;
        y -= this.cy;
        return [
            (this.cos*x + this.sin*y) / this.rx,
            (this.cos*y - this.sin*x) / this.ry
        ];
    }

    /**
     * Map a unit circle point to a point on this ellipse.
     */
    invert(x: number | XY, y?: number): XY {
        if (x instanceof Array) {
            y = x[1];
            x = x[0];
        } else if (y === undefined) {
            throw new Error("expected y in Ellipse.invert: " + JSON.stringify(x));
        }
        const rX = this.rx * x;
        const rY = this.ry * y;
        const X = this.cos * rX - this.sin * rY;
        const Y = this.cos * rY + this.sin * rX;
        return [
            X + this.cx,
            Y + this.cy
        ];
    }

    translate(tx: number, ty: number): Ellipse {
        const {A, B, C, D, E, F, theta, color, idx, name} = this;
        return new Ellipse({
            A: A,
            B: B,
            C: C,
            D: D - 2 * A * tx - B * ty,
            E: E - 2 * C * ty - B * tx,
            F: F + A * tx * tx + B * tx * ty + C * ty * ty - D * tx - E * ty,
            theta, color, idx, name,
        });
    }

    /**
     * Rotate this ellipse around the origin (0,0)
     */
    rotate(t: number): Ellipse {
        const {A, B, C, D, E, F, theta, color, idx, name,} = this;
        const c = Math.cos(t);
        const s = Math.sin(t);
        const c2 = c * c;
        const s2 = s * s;
        const cs = c * s;
        return new Ellipse({
            A: A*c2 - B*cs + C*s2,
            B: 2*cs*(A-C) + B*(c2 - s2),
            C: C*c2 + B*cs + A*s2,
            D: D*c - E*s,
            E: D*s + E*c,
            F: F,
            theta: theta + t,
            color, idx, name,
        });
    }

    /**
     * Scale this ellipse by { sx, sy }.
     */
    scale(sx: number, sy: number): Ellipse {
        const {A, B, C, D, E, F, theta, color, idx, name} = this;
        return new Ellipse({
            A: A / sx / sx,
            B: B / sx / sy,
            C: C / sy / sy,
            D: D / sx,
            E: E / sy,
            F: F,
            theta, color, idx, name,
        });
    }

    /**
     * Apply a linear transform to this ellipse.
     */
    affine(sx: number, sy: number, theta: number, tx: number, ty: number): Ellipse {
        return this.translate(tx, ty).rotate(theta).scale(sx, sy);
    }

    /**
     * Project this ellipse into a plane where the provided ellipse is a unit circle.
     */
    project(e: Ellipse): Ellipse {
        return this.affine(1/e.rx, 1/e.ry, -e.theta, -e.cx, -e.cy);
    }

    /**
     * Get a point on the ellipse at angle `degrees`°
     */
    getDegrees(degrees: number): XY {
        return this.getPoint(degrees * pi / 180);
    }

    /**
     * Get a point on the ellipse at angle `theta` (radians)
     */
    getPoint(theta: number): XY {
        return this.invert(Math.cos(theta), Math.sin(theta));
    }

    containsEllipse(e: Ellipse) {
        return this.contains(e.vx1) && this.contains(e.vx2) && this.contains(e.vy1) && this.contains(e.vy2);
    }

    contains(px: number | Point | XY, py?: number) {
        if (px instanceof Array) {
            py = px[1];
            px = px[0];
        } else if (typeof px === 'object' && 'x' in px) {
            py = px.y;
            px = px.x;
        } else if (py === undefined) {
            throw new Error("expected y in Ellipse.contains: " + JSON.stringify(px));
        }
        const [x, y] = this.transform(px, py);
        const r2 = x * x + y * y;
        return r2 <= 1;
    }

    intersect(e: Ellipse): Intersection[] {
        const e1 = this;
        const e2 = e;
        const p1 = e1.project(e2);
        const uis = p1.unitIntersections();

        return uis.map(ui => {
            const [c2, s2] = ui;
            const [x, y] = e2.invert(c2, s2);
            return new Intersection({ e1, e2, x, y, c2, s2 });
        });
    }

    unitIntersections(): XY[] {
        const {A, B, C, D, E, F} = this;

        const B2 = B * B;
        const E2 = E * E;
        const BE = B * E;
        const AC = A - C;
        const CF = C + F;

        let c4 = AC * AC + B2;
        let c3 = 2 * D * AC + 2 * BE;
        let c2 = D * D + 2 * AC * CF + E2 - B2;
        let c1 = 2 * D * CF - 2 * BE;
        let c0 = CF * CF - E2;

        [c4, c3, c2, c1, c0] = [c4, c3, c2, c1, c0].map(zeroCheck);

        //console.log("coeffs:", c4,c3,c2,c1,c0);

        const xs = quartic(c4, c3, c2, c1, c0);
        //console.log("quartic:", xs);
        const xo = new Map<number, number>();
        xs.forEach((x: number) => { xo.set(x, (xo.get(x) || 0) + 1); })

        //console.log("xo:", xo);

        //var rxs = [];
        //var ys = [];
        const ps: XY[] = [];
        xo.forEach((n, x) => {
            let y = sqrt(1 - x * x);
            if (isNaN(y)) return;
            const b = A * x * x + C * y * y + D * x + F;
            const c = B * x * y + E * y;
            const r1 = Math.abs(b + c);
            const r2 = Math.abs(b - c);

            //console.log("x:", x, "n:", n, "rs:", r1, r2, r1 - r2);
            if (n > 1) {
                //console.log("pushing double:", y, -y);
                ps.push([x, -y]);
                ps.push([x, y]);
            } else if (r2 < r1) {
                ps.push([x, -y]);
            } else {
                ps.push([x, y]);
            }
            //console.log("\t", pp(x,r1),pp(x,r2));
        });

        //console.log("unit:", this.toString(), "coeffs:", [c4,c3,c2,c1,c0].map(r3), "xs:", xs.map(r3), "ps:", pps(ps));
        return ps;
    }
}
